src/http/modules/ngx_http_upstream_zone_module.c

```diff
@@ -124,26 +126,81 @@ ngx_http_upstream_zone(ngx_conf_t *cf, ngx_command_t *cmd, void *conf)
     uscf->shm_zone->init = ngx_http_upstream_init_zone;
     uscf->shm_zone->data = umcf;
 
-    uscf->shm_zone->noreuse = 1;
-
     return NGX_CONF_OK;
 }
 
 
 static ngx_int_t
 ngx_http_upstream_init_zone(ngx_shm_zone_t *shm_zone, void *data)
 {
+    ngx_http_upstream_main_conf_t  *oumcf = data;
+
     size_t                          len;
     ngx_uint_t                      i, j;
     ngx_slab_pool_t                *shpool;
     ngx_http_upstream_rr_peers_t   *peers, **peersp;
     ngx_http_upstream_srv_conf_t   *uscf, *ouscf, **uscfp, **ouscfp;
-    ngx_http_upstream_main_conf_t  *umcf, *oumcf;
+    ngx_http_upstream_main_conf_t  *umcf;
 
     shpool = (ngx_slab_pool_t *) shm_zone->shm.addr;
     umcf = shm_zone->data;
     uscfp = umcf->upstreams.elts;
 
+    if (oumcf && shpool->data) {
+
+        for (i = 0; i < umcf->upstreams.nelts; i++) {
+            uscf = uscfp[i];
+
+            if (uscf->shm_zone != shm_zone) {
+                continue;
+            }
+
+            for (j = 0; j < oumcf->upstreams.nelts; j++) {
+                ouscfp = oumcf->upstreams.elts;
+
+                 if (ouscfp[j]->shm_zone == NULL) {
+                     continue;
+                 }
+
+                 if (ouscfp[j]->shm_zone->shm.name.len != shm_zone->shm.name.len
+                     || ngx_memcmp(ouscfp[j]->shm_zone->shm.name.data,
+                                   shm_zone->shm.name.data,
+                                   shm_zone->shm.name.len)
+                        != 0)
+                 {
+                     continue;
+                 }
+
+                 if (ngx_http_upstream_zone_servers_equal(uscf, ouscfp[j])
```

Comment:
This looks suspiciously straightforward. Is it OK to reuse zone with different LB algorithm?